pydantic_ai_slim/pydantic_ai/mcp.py

```diff
@@ -110,8 +110,7 @@ async def list_tools(self) -> list[mcp_types.Tool]:
         - We don't cache tools as they might change.
         - We also don't subscribe to the server to avoid complexity.
         """
-        async with self:  # Ensure server is running
```

Comment:
I still want this in case the MCP server is being used outside of agent context, and in Temporal where each list/call tool call happens inside a separate activity: https://github.com/pydantic/pydantic-ai/pull/2225

Reply:
In the description you write:

> Always enter the Agent's context in .iter -- this allows us to remove the need for the MCP Toolset to enter its context in get_tools -- which I believe was resulting in startup/shutdown for every get_tools call.

But I think it's OK to enter too many times as we keep track of the running count, it shouldn't actually shutdown the server on each get_tools call if we've already entered the toolsets/server in a higher context.

Reply:
Yes, I was only talking about the agent case, if the context is not entered at all I think you get the startup shutdown behavior, entering too many times is a-ok